File: 문제/비트마스킹/19942.cpp
#include <bits/stdc++.h>
using namespace std;

const int INF = 987654321;

int N;
int mp, mf, ms, mv;
int cp, cf, cs, cv, cost;
int sum;
int ret = INF;

struct Foods {
  int mp, mf, ms, mv, cost;
} foods[16];

map<int, vector<vector<int>>> retV;

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  cout.tie(NULL);

  cin >> N;
  cin >> mp >> mf >> ms >> mv;

  for (int i = 1; i <= N; i++) {
    cin >> cp >> cf >> cs >> cv >> cost;

    foods[i].mp = cp;
    foods[i].mf = cf;
    foods[i].ms = cs;
    foods[i].mv = cv;
    foods[i].cost = cost;
  }

  for (int i = 1; i < (1 << N); i++) {
    vector<int> v;
    cp = cf = cs = cv = sum = 0;

    for (int j = 0; j < N; j++) {
      if (i & (1 << j)) {
        int idx = j + 1;

        v.push_back(idx);
        cp += foods[idx].mp;
        cf += foods[idx].mf;
        cs += foods[idx].ms;
        cv += foods[idx].mv;
        sum += foods[idx].cost;
      }
    }

    if (cp >= mp && cf >= mf && cs >= ms && cv >= mv) {
      if (sum <= ret) {
        ret = sum;
        retV[ret].push_back(v);
      }
    }
  }

  if (ret == INF)
    cout << -1 << "\n";
  else {
    sort(retV[ret].begin(), retV[ret].end());
    cout << ret << "\n";

    for (auto v : retV[ret][0]) {
      cout << v << " ";
    }
  }

  return 0;
}

File: 문제/비트마스킹/19942.js
const fs = require("fs");
const stdin = (
  process.platform === "linux"
    ? fs.readFileSync("/dev/stdin").toString()
    : `6
100 70 90 10
30 55 10 8 100
60 10 10 2 70
10 80 50 0 50
40 30 30 8 60
60 10 70 2 120
20 70 50 4 4`
).split("\n");

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

function solution() {
  const N = Number(input());
  const [mp, mf, ms, mv] = input().split(" ").map(Number);
  const foods = {};
  let ret = Number.MAX_SAFE_INTEGER;
  let retArr = [];

  for (let i = 1; i <= N; i++) {
    const [p, f, s, v, c] = input().split(" ").map(Number);

    foods[i] = {
      p,
      f,
      s,
      v,
      c,
    };
  }

  for (let i = 1; i < 1 << N; i++) {
    const selected = [];
    let [p, f, s, v, sum] = [0, 0, 0, 0, 0];

    for (let j = 0; j < N; j++) {
      if (i & (1 << j)) {
        const idx = j + 1;
        selected.push(idx);
        p += foods[idx].p;
        f += foods[idx].f;
        s += foods[idx].s;
        v += foods[idx].v;
        sum += foods[idx].c;
      }
    }

    if (p >= mp && f >= mf && s >= ms && v >= mv) {
      if (ret >= sum) {
        if (ret > sum) retArr = [];

        ret = sum;
        retArr.push(selected.join(" "));
      }
    }
  }

  if (ret === Number.MAX_SAFE_INTEGER) {
    return -1;
  } else {
    retArr.sort();
    return [ret, retArr[0]].join("\n");
  }
}

console.log(solution());
